File: library-management/csharp/src/LibraryManagement/Copy.cs
namespace LibraryManagement;

public class Copy
{
    public Copy(int id, Isbn isbn)
    {
        Id = id;
        Isbn = isbn;
        Status = CopyStatus.Available;
    }

    public int Id { get; }
    public Isbn Isbn { get; }
    public CopyStatus Status { get; private set; }

    internal void MarkCheckedOut() => Status = CopyStatus.CheckedOut;
    internal void MarkAvailable() => Status = CopyStatus.Available;
    internal void MarkReserved() => Status = CopyStatus.Reserved;
}

File: library-management/typescript/src/Copy.ts
import { CopyStatus } from './CopyStatus.js';
import { Isbn } from './Isbn.js';

export class Copy {
  private _status: CopyStatus = CopyStatus.Available;

  constructor(public readonly id: number, public readonly isbn: Isbn) {}

  get status(): CopyStatus { return this._status; }

  markCheckedOut(): void { this._status = CopyStatus.CheckedOut; }
  markAvailable(): void { this._status = CopyStatus.Available; }
  markReserved(): void { this._status = CopyStatus.Reserved; }
}
